docker/webapp/app.py

```diff
@@ -0,0 +1,258 @@
+# TODO:
+#  - rss
+#  - opensearch
+#  - unify static files, no dup melpa.css etc
+#  - toggle stable vs. non-stable, add link at top-right to the other
+#  - build timestamps
+#  - page titles
+#  - entity-quote package URLs containing +
+#  - case-insensitive search
+#  - full URLs for badges
+#  - gist recipes
+#  - escaping chars in description etc.
+#  - serve packages statically by default, use /packages/foo URLs.
+
+from fastapi import FastAPI, Request, HTTPException
+from fastapi.exceptions import RequestValidationError
+from fastapi.responses import HTMLResponse, PlainTextResponse, RedirectResponse
+from fastapi.staticfiles import StaticFiles
+from fastapi.templating import Jinja2Templates
+from starlette.datastructures import URL
+import starlette.exceptions
+import json
+import os
+import datetime
+from collections import namedtuple
+import re
+from itertools import groupby
+from math import ceil
+
+##############################################################################
+# Data models
+##############################################################################
+
+HTML_DIR="../../html"
+
+BuildStatus = namedtuple("BuildStatus", ['started_at', 'completed_at', 'next_at', 'duration'])
+
+def calculate_source_url(recipe, commit):
+      fetcher = recipe["fetcher"]
+      url = recipe.get("url")
+      repo = recipe.get("repo")
+      branch = recipe.get("branch")
+      match fetcher:
+          case 'github':
+              ref = commit or branch
+              extra = ref and f"/tree/{ref}" or ""
+              return f"https://github.com/{repo}{extra}"
+          case "gitlab":
+              ref = commit or branch
+              extra = ref and f"/tree/{ref}" or ""
+              return f"https://gitlab.com/{repo}{extra}"
+          case "sourcehut":
+              ref = commit or branch
+              extra = ref and f"/tree/{ref}" or ""
+              return f"https://git.sr.ht/~{repo}{extra}"
+          case "bitbucket":
+              extra = ""
+              if commit: extra = "/src/" + commit
+              elif branch: extra = "/branch/" + branch
+              return f"https://bitbucket.com/{repo}{extra}"
+          case _:
+              if not url: return None
+              def url_match(pat, rep=''):
+                  new_url = re.sub(pat, rep, url)
+                  if new_url != url: return new_url
+                  else: return None
+              return (url_match(r'(bitbucket\.org\/[^\/]+\/[^\/\?]+)', "https://") or
```

Comment:
Maybe use [`urllib.parse.urlparse()`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse) instead?

Reply:
That would be a good refactoring. This was just a transliteration of the javascript code that's been working for years.

Reply:
On second thoughts, the result would be _much_ more verbose for what I'm doing here.